commit 84ee79703d81758e581ea1849d2140b3474b4192
Author: Yonghua Zheng <yonghua.zheng@esimtek.com>
Date:   Fri Jun 9 18:20:32 2017 +0800

    FIX mixed-using of control interface between global and iface
    
    Android KK Wifi framework has changed to use global interface
    however revelant changes in wpa_supplicant are incomplete: both
    global and iface use the same 'wpa_wlan0' as unix socket path
    to create socket and register callback to poll, as both use
    the same fd, callbacks in poll will be invoked and executed
    and could cause unexpected results.
    
    This patch provides a workaround and prevents iface interface
    from registering 'poll' callback to the same fd created by
    global interface, if it is enabled and use the same socket path.
    
    Change-Id: Id25f2fc8904b5d3c92f860d5031f86cf47663078

diff --git a/wpa_supplicant/ctrl_iface_unix.c b/wpa_supplicant/ctrl_iface_unix.c
index e35d2c3..50029f6 100644
--- a/wpa_supplicant/ctrl_iface_unix.c
+++ b/wpa_supplicant/ctrl_iface_unix.c
@@ -455,6 +455,52 @@ wpa_supplicant_ctrl_iface_init(struct wpa_supplicant *wpa_s)
 
 #ifdef ANDROID
 havesock:
+	if (wpa_s->global->params.ctrl_interface) {
+		int same = 0;
+
+		if (wpa_s->global->params.ctrl_interface[0] == '/') {
+			if (os_strcmp(wpa_s->global->params.ctrl_interface,
+						wpa_s->conf->ctrl_interface) == 0)
+				same = 1;
+		} else if (os_strncmp(wpa_s->global->params.ctrl_interface,
+					"@android:", 9) == 0 ||
+				os_strncmp(wpa_s->global->params.ctrl_interface,
+					"@abstract:", 10) == 0) {
+			char *pos;
+
+			/*
+			 * Currently, Android uses @android:wpa_* as the naming
+			 * convention for the global ctrl interface. This logic
+			 * needs to be revisited if the above naming convention
+			 * is modified.
+			 */
+			pos = os_strchr(wpa_s->global->params.ctrl_interface,
+					'_');
+			if (pos &&
+				os_strcmp(pos + 1,
+					wpa_s->conf->ctrl_interface) == 0)
+				same = 1;
+		}
+
+		if (same) {
+			/*
+			 * The invalid configuration combination might be
+			 * possible to hit in an Android OTA upgrade case, so
+			 * instead of refusing to start the wpa_supplicant
+			 * process, do not open the per-interface ctrl_iface
+			 * and continue with the global control interface that
+			 * was set from the command line since the Wi-Fi
+			 * framework will use it for operations.
+			 */
+			wpa_printf(MSG_ERROR,
+					"global ctrl interface %s matches ctrl interface %s - do not open per-interface ctrl interface",
+					wpa_s->global->params.ctrl_interface,
+					wpa_s->conf->ctrl_interface);
+			priv->sock = dup(priv->sock);
+			wpa_msg_register_cb(wpa_supplicant_ctrl_iface_msg_cb);
+			return priv;
+		}
+	}
 #endif /* ANDROID */
 
 	/*
